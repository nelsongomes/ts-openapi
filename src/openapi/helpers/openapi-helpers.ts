import {
  Body,
  SecurityApiKeyScheme,
  SecurityApiKeySchemeMethod,
  SecurityBasicScheme,
  SecurityBearerScheme,
} from "../openapi.types";

export function textPlain(description: string): Body {
  return {
    description,
    content: { "text-plain": {} },
  };
}

export function basicAuth(): SecurityBasicScheme {
  return {
    type: "http",
    scheme: "basic",
  };
}

export function apiKeyAuth(
  name: string,
  method: SecurityApiKeySchemeMethod
): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: method,
    name,
  };
}

export function cookieAuth(cookieName: string): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: "cookie",
    name: cookieName,
  };
}

export function bearerAuth(
  bearerFormat?: "JWT" | string
): SecurityBearerScheme {
  return {
    type: "http",
    scheme: "bearer",
    ...(bearerFormat && { bearerFormat }),
  };
}
