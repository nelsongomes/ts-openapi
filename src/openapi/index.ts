import Joi, { Schema } from "joi";

export { Joi, Schema };

export { OpenApi } from "./openapi";
export { WebRequestSchema, ParameterIn, Parameters } from "./openapi.types";

export { textPlain } from "./openapi-helpers";
