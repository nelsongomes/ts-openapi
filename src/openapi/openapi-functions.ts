import Joi, { ObjectSchema } from "joi";
import joiToSwagger from "../joi-conversion";
import {
  Body,
  IntegerFormats,
  NumberFormats,
  SchemaTypeArray,
  SchemaTypeBoolean,
  SchemaTypeInteger,
  SchemaTypeNumber,
  SchemaTypeObject,
  SchemaTypeString,
  StringFormats
} from "./openapi.types";

export function limitations(parameter: any): string {
  const limitationArray = [];

  if (parameter.type === "string" && parameter.enum) {
    // min & max should be ignored when is an enum
    delete parameter.minLength;
    delete parameter.maxLength;
  }

  // string
  if (parameter.minLength) {
    limitationArray.push(`min:${parameter.minLength} chars`);
  }

  if (parameter.maxLength) {
    limitationArray.push(`max:${parameter.maxLength} chars`);
  }

  // integer / float
  if (parameter.type === "integer" && parameter.enum) {
    // min & max should be ignored when is an enum
    delete parameter.minimum;
    delete parameter.maximum;
  }

  if (typeof parameter.minimum === "number") {
    limitationArray.push(`min:${parameter.minimum}`);
  }

  if (typeof parameter.maximum === "number") {
    limitationArray.push(`max:${parameter.maximum}`);
  }

  // array
  if (parameter.minItems) {
    limitationArray.push(`minItems:${parameter.minItems}`);
  }

  if (parameter.maxItems) {
    limitationArray.push(`maxItems:${parameter.maxItems}`);
  }

  if (parameter.format) {
    switch (parameter.format) {
      case StringFormats.DateTime:
        if (parameter.meta && parameter.meta.format === "date") {
          limitationArray.push(`date:yyyy-mm-dd`);
        } else {
          limitationArray.push(`ISO8601 date-time format`);
        }
        break;
      case StringFormats.Byte:
        limitationArray.push("base64 encoded string");
        break;
      case StringFormats.Binary:
        limitationArray.push("binary string");
        break;
    }
  }

  return limitationArray.length > 0 ? ` (${limitationArray.join(", ")})` : "";
}

export function bodyParams(schema: ObjectSchema): Body {
  const internalSchema = Joi.object().keys({ object: schema.required() });
  const query = joiToSwagger(internalSchema, {});

  const key = Object.keys(query.swagger.properties)[0];
  const parameter = query.swagger.properties[key];

  return {
    description:
      query.swagger.properties[key].description ||
      "Body does not have a description.",
    content: {
      "application/json": {
        schema: objectSchema(parameter),
        ...(parameter.example && { example: parameter.example })
      }
    }
  };
}

function objectSchema(parameter: any): SchemaTypeObject {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeObject = {
    description,
    ...(typeof parameter.default === "object" && {
      default: parameter.default
    }),
    ...(parameter.minItems && { minItems: parameter.minItems }),
    ...(parameter.maxItems && { maxItems: parameter.maxItems }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "object",
    properties: {}
  };

  for (const propertyKey of Object.keys(parameter.properties)) {
    const property = parameter.properties[propertyKey];

    switch (property.type) {
      case "string":
        const {
          default: ignoreA1,
          nullable: ignoreA2,
          minLength,
          maxLength,
          ...restOfStringSchema
        } = stringSchema(property);

        output.properties[propertyKey] = restOfStringSchema;
        break;
      case "number":
        const {
          default: ignoreB1,
          nullable: ignoreB2,
          minimum: ignoreB3,
          maximum: ignoreB4,
          ...restNumberOfSchema
        } = numberSchema(property);

        output.properties[propertyKey] = restNumberOfSchema;
        break;
      case "boolean":
        const {
          default: ignoreE1,
          nullable: ignorE2,
          ...restBooleanOfSchema
        } = booleanSchema(property);

        output.properties[propertyKey] = restBooleanOfSchema;
        break;
      case "integer":
        const {
          default: ignoreC1,
          nullable: ignoreC2,
          minimum: ignoreC3,
          maximum: ignoreC4,
          ...restIntegerOfSchema
        } = integerSchema(property);

        output.properties[propertyKey] = restIntegerOfSchema;
        break;
      case "object":
        const {
          default: ignoreD1,
          nullable: ignoreD2,
          ...restOfObjectSchema
        } = objectSchema(property);

        output.properties[propertyKey] = restOfObjectSchema;
        break;
      case "array":
        const {
          default: ignoreF1,
          nullable: ignoreF2,
          ...restOfArraySchema
        } = arraySchema(property);

        output.properties[propertyKey] = restOfArraySchema;
        break;
      default:
        throw new Error(`${property.type} not implemented`);
    }
  }

  return output;
}

export function stringSchema(parameter: any): SchemaTypeString {
  const supportedFormats = Object.values(StringFormats);
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeString = {
    description,
    ...(parameter.default && { default: parameter.default }),
    ...(parameter.format &&
      supportedFormats.includes(parameter.format) && {
        format: parameter.format
      }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(parameter.minLength && { minLength: parameter.minLength }),
    ...(parameter.maxLength && { maxLength: parameter.maxLength }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "string"
  };

  if (parameter.meta) {
    if (parameter.meta.format) {
      parameter.format = parameter.meta.format;
    }
  }

  if (parameter.format && supportedFormats.includes(parameter.format)) {
    output.format = parameter.format;

    switch (parameter.format) {
      case StringFormats.Date:
        output.maxLength = 10; // 2020-10-14
        output.minLength = 10;
        break;
      case StringFormats.DateTime:
        output.minLength = 16; // 20201014T214403Z
        output.maxLength = 25; // 2020-10-14T21:44:03+00:00
        break;
      case StringFormats.Password:
        delete output.default; // no defaults for passwords
        break;
    }
  }

  // default values must be part of enum
  if (
    parameter.enum instanceof Array &&
    !parameter.enum.includes(output.default)
  ) {
    delete output.default;
  }

  return output;
}

export function numberSchema(parameter: any): SchemaTypeNumber {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeNumber = {
    description,
    ...(parameter.default && { default: parameter.default }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(typeof parameter.minimum === "number" && {
      minimum: parameter.minimum
    }),
    ...(typeof parameter.maximum === "number" && {
      maximum: parameter.maximum
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    format: NumberFormats.Double,
    type: "number"
  };

  return output;
}

export function booleanSchema(parameter: any): SchemaTypeBoolean {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeBoolean = {
    description,
    ...(typeof parameter.default === "boolean" && {
      default: parameter.default
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "boolean"
  };

  return output;
}

export function integerSchema(parameter: any): SchemaTypeInteger {
  const description =
    (parameter.description || "Parameter without description.") +
    (limitations(parameter) || "");

  const output: SchemaTypeInteger = {
    description,
    ...(typeof parameter.default === "number" && {
      default: parameter.default
    }),
    ...(typeof parameter.example === "number" && {
      example: parameter.example
    }),
    ...(parameter.enum && { enum: parameter.enum }),
    ...(typeof parameter.minimum === "number" && {
      minimum: parameter.minimum
    }),
    ...(typeof parameter.maximum === "number" && {
      maximum: parameter.maximum
    }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    format: IntegerFormats.Int64,
    type: "integer"
  };

  // default values must be part of enum
  if (
    parameter.enum instanceof Array &&
    !parameter.enum.includes(output.default)
  ) {
    delete output.default;
  }

  return output;
}

export function arraySchema(parameter: any): SchemaTypeArray {
  const output: SchemaTypeArray = {
    ...(typeof parameter.default === "object" && {
      default: parameter.default
    }),
    ...(parameter.minItems && { minItems: parameter.minItems }),
    ...(parameter.maxItems && { maxItems: parameter.maxItems }),
    ...(typeof parameter.nullable === "boolean" && {
      nullable: parameter.nullable
    }),
    type: "array"
  };

  switch (parameter.items.type) {
    case "string":
      const {
        default: ignoreA1,
        nullable: ignoreA2,
        minLength,
        maxLength,
        ...restOfStringSchema
      } = stringSchema(parameter.items);

      output.items = restOfStringSchema;
      break;
    case "number":
      const {
        default: ignoreB1,
        nullable: ignoreB2,
        minimum: ignoreB3,
        maximum: ignoreB4,
        ...restNumberOfSchema
      } = numberSchema(parameter.items);

      output.items = restNumberOfSchema;
      break;
    case "integer":
      const {
        default: ignoreC1,
        nullable: ignoreC2,
        minimum: ignoreC3,
        maximum: ignoreC4,
        ...restIntegerOfSchema
      } = integerSchema(parameter.items);

      output.items = restIntegerOfSchema;
      break;
    case "object":
      const {
        default: ignoreD1,
        nullable: ignoreD2,
        ...restOfObjectSchema
      } = objectSchema(parameter.items);

      output.items = restOfObjectSchema;
      break;
    default:
      throw new Error("not implemented");
  }

  if (parameter.items.enum instanceof Array) {
    // default values must be part of enum
    if (
      output.default instanceof Array &&
      !output.default.every((element: any, _index: number, _array: any[]) =>
        parameter.items.enum.includes(element)
      )
    ) {
      delete output.default;
    }

    // string enums are already limited
    if (parameter.items.type === "string") {
      delete (output.items as SchemaTypeString).minLength;
      delete (output.items as SchemaTypeString).maxLength;
    }
  }

  return output;
}

export function bodySchema(schema: ObjectSchema): Body {
  return bodyParams(schema);
}
