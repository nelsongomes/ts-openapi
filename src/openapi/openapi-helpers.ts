export function textPlain(description: string) {
  return {
    content: {
      "text/plain": {
        schema: {
          type: "string",
        },
      },
    },
    description,
  };
}
