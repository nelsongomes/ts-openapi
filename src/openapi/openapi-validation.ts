import { ParameterIn } from "./openapi.types";

export function validateParameters(
  type: ParameterIn,
  parameter: any,
  name: string,
  parameterType: string,
  isParameterRequired: boolean
) {
  switch (type) {
    case ParameterIn.Path:
      validatePathParameters(name, parameterType, isParameterRequired);
      break;
    case ParameterIn.Query:
      validateQueryParameters(
        parameter,
        name,
        parameterType,
        isParameterRequired
      );
      break;
    case ParameterIn.Cookie:
    case ParameterIn.Header:
      const upcaseType = type.charAt(0).toUpperCase() + type.slice(1);
      validateCookieParameters(upcaseType, name, parameterType);
      break;
  }
}

function validatePathParameters(
  key: string,
  parameterType: string,
  isParameterRequired: boolean
) {
  const nameMatch = key.match(/[A-Za-z0-9_]+/);

  if (nameMatch && nameMatch[0] !== key) {
    throw new Error(`Path param '${key}' name does not match [A-Za-z0-9_]`);
  }

  if (["object", "array"].includes(parameterType)) {
    throw new Error(`Path param '${key}' cannot be an object or an array.`);
  }

  if (!isParameterRequired) {
    throw new Error(
      `Path param '${key}' must be required because it is a path parameter.`
    );
  }
}

function validateQueryParameters(
  parameter: any,
  key: string,
  parameterType: string,
  _isParameterRequired: boolean
) {
  if (["object"].includes(parameterType)) {
    throw new Error(`Query param '${key}' cannot be an object.`);
  }

  if (["array"].includes(parameterType) && parameter.items.properties) {
    const arrayProperties = parameter.items.properties;

    for (const propKey of Object.keys(arrayProperties)) {
      const property = arrayProperties[propKey];

      if (property.type === "array") {
        throw new Error(
          `Query param '${key}' type array can only have scalar types inside of it (cannot be an array of arrays or an array of objects).`
        );
      }
    }
  }
}

function validateCookieParameters(
  type: string,
  key: string,
  parameterType: string
) {
  if (["object", "array"].includes(parameterType)) {
    throw new Error(`${type} param '${key}' cannot be an object or an array.`);
  }
}
