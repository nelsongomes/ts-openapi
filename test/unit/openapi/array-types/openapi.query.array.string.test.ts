import Joi, { number } from "joi";
import { example } from "yargs";
import { OpenApi } from "../../../../src/openapi/openapi";
import { textPlain } from "../../../../src/openapi/openapi-helpers";
import {
  Parameters,
  ParameterIn,
  WebRequestSchema,
} from "../../../../src/openapi/openapi.types";

describe("src/openapi/openapi", () => {
  let openApi: OpenApi;

  describe("Array of string", () => {
    beforeEach(() => {
      openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      openApi.setServers([{ url: "https://server.com" }]);
      openApi.addPath(
        "/health",
        {
          get: {
            description: "Service healthcheck endpoint",
            operationId: "repeated",
            parameters: [],
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Healthcheck",
            tags: ["Internals"],
          },
        },
        true
      );
    });

    test("string array simple", async () => {
      const parameters: Parameters = [];

      const query = Joi.object().keys({
        category: Joi.array().items(Joi.string()),
      });

      openApi.genericParams(parameters, query, ParameterIn.Query);
      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            parameters,
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Test",
            tags: ["Internals"],
          },
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("string array all options", async () => {
      const parameters: Parameters = [];
      const query = Joi.object({
        category: Joi.array()
          .items(
            Joi.string()
              // ignore these
              .min(11)
              .max(20)
              .required()
              .example("example")
              .default("default")
              .allow(null)
          )
          .min(1)
          .max(50)
          .description("String array")
          .required()
          .example(["a"])
          .default("b")
          .allow(null),
      });

      openApi.genericParams(parameters, query, ParameterIn.Query);
      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            parameters,
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Test",
            tags: ["Internals"],
          },
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("string array enum", async () => {
      enum EnumValues {
        AAA = "a",
        BBB = "b",
        CCC = "c",
      }

      const parameters: Parameters = [];
      const query = Joi.object({
        category: Joi.array().items(
          Joi.string().valid(...Object.values(EnumValues))
        ),
      });

      openApi.genericParams(parameters, query, ParameterIn.Query);
      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            parameters,
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Test",
            tags: ["Internals"],
          },
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("string array enum min and max should be skipped (the values are already limited)", async () => {
      enum EnumValues {
        AAA = "a",
        BBB = "b",
        CCC = "c",
      }

      const parameters: Parameters = [];
      const query = Joi.object().keys({
        category: Joi.array()
          .items(
            Joi.string()
              .valid(...Object.values(EnumValues))
              .max(1)
              .min(10)
          )
          .required(),
      });

      openApi.genericParams(parameters, query, ParameterIn.Query);
      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            parameters,
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Test",
            tags: ["Internals"],
          },
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("string array enum default is part of enum", async () => {
      enum EnumValues {
        AAA = "a",
        BBB = "b",
        CCC = "c",
      }

      const parameters: Parameters = [];
      const query = Joi.object().keys({
        category: Joi.array()
          .items(Joi.string().valid(...Object.values(EnumValues)))
          .default(["a", "b"]),
      });

      openApi.genericParams(parameters, query, ParameterIn.Query);
      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            parameters,
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Test",
            tags: ["Internals"],
          },
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("string array enum default is undefined because is NOT in enum", async () => {
      enum EnumValues {
        AAA = "a",
        BBB = "b",
        CCC = "c",
      }

      const parameters: Parameters = [];
      const query = Joi.object().keys({
        category: Joi.array()
          .items(Joi.string().valid(...Object.values(EnumValues)))
          .default(["a", "z"]),
      });

      openApi.genericParams(parameters, query, ParameterIn.Query);
      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            parameters,
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Test",
            tags: ["Internals"],
          },
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("string array enum all options", async () => {
      enum EnumValues {
        AAA = "a",
        BBB = "b",
        CCC = "c",
      }

      const parameters: Parameters = [];
      const query = Joi.object({
        category: Joi.array()
          .items(
            Joi.string()
              .valid(...Object.values(EnumValues))
              // ignore these
              .description("Just a string")
              .min(10)
              .max(20)
              .example("ignore")
              .default("ignore")
              .required()
              .allow(null)
          )
          .description("array of category")
          .required()
          .min(1)
          .max(10)
          .example(["a", "b", "c"])
          .default(["b"])
          .allow(null),
      }).description("ignore this");

      openApi.genericParams(parameters, query, ParameterIn.Query);
      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            parameters,
            responses: {
              200: textPlain("Successful operation."),
            },
            summary: "Server Test",
            tags: ["Internals"],
          },
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });
  });
});
