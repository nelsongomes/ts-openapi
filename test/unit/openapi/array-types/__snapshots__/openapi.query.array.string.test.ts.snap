// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/openapi/openapi Array of string string array all options 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "String array (minItems:1, maxItems:50)",
            "example": Array [
              "a",
            ],
            "in": "query",
            "name": "category",
            "required": true,
            "schema": Object {
              "items": Object {
                "type": "string",
              },
              "maxItems": 50,
              "minItems": 1,
              "nullable": true,
              "type": "array",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Array of string string array enum 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Parameter without description.",
            "in": "query",
            "name": "category",
            "required": false,
            "schema": Object {
              "items": Object {
                "enum": Array [
                  "a",
                  "b",
                  "c",
                ],
                "type": "string",
              },
              "type": "array",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Array of string string array enum all options 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "array of category (minItems:1, maxItems:10)",
            "example": Array [
              "a",
              "b",
              "c",
            ],
            "in": "query",
            "name": "category",
            "required": true,
            "schema": Object {
              "default": Array [
                "b",
              ],
              "items": Object {
                "enum": Array [
                  "a",
                  "b",
                  "c",
                ],
                "type": "string",
              },
              "maxItems": 10,
              "minItems": 1,
              "nullable": true,
              "type": "array",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Array of string string array enum default is part of enum 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Parameter without description.",
            "in": "query",
            "name": "category",
            "required": false,
            "schema": Object {
              "default": Array [
                "a",
                "b",
              ],
              "items": Object {
                "enum": Array [
                  "a",
                  "b",
                  "c",
                ],
                "type": "string",
              },
              "type": "array",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Array of string string array enum default is undefined because is NOT in enum 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Parameter without description.",
            "in": "query",
            "name": "category",
            "required": false,
            "schema": Object {
              "items": Object {
                "enum": Array [
                  "a",
                  "b",
                  "c",
                ],
                "type": "string",
              },
              "type": "array",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Array of string string array enum min and max should be skipped (the values are already limited) 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Parameter without description.",
            "in": "query",
            "name": "category",
            "required": true,
            "schema": Object {
              "items": Object {
                "enum": Array [
                  "a",
                  "b",
                  "c",
                ],
                "type": "string",
              },
              "type": "array",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Array of string string array simple 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Parameter without description.",
            "in": "query",
            "name": "category",
            "required": false,
            "schema": Object {
              "items": Object {
                "type": "string",
              },
              "type": "array",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;
